blog article about creating wrapper scripts for DDEV

agviu opened the pull request:
As promised, I created a blog article about how we use DDEV in my current company.
Please feel free to add any head image or whatever you consider necessary to make it look as other blog posts.

rfay commented:
Yay!

rfay reviewed (commented):
I'll have some usage things to suggest after you get tests passing, but the first question that people are going to ask is...

What is your wrapper script and what does it do? Should I use it?

rfay commented:
I understand that it's not publicly available, but explaining the basics of what you did in order to smooth thing over would help people know whether it's a worthy thing to do. And I guess it's been successful over time? You have to maintain separate docs? Do people understand how to use the underlying DDEV functionality?

agviu commented:
Hi @rfay , thanks for your feedback and suggestions.
I have updated the article trying to cover everything you asked and a bit more.

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+- Make the tool as easy to use as possible, with an independent documentation. Starting with a 'help' command that can describe all the other commands and the options that are accepted.
+- Built with latests PHP features, using Symfony project when needed (yaml...), PHPUnit...
+- Use Phar format for distributing the lsm tool, which also has its own version lifecycle and self-update options.
+- Possibility of having several instances installed at the same time without any conflict.
```
```suggestion
- Allow having several versions of the script installed at the same time without conflict.
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+- Use Phar format for distributing the lsm tool, which also has its own version lifecycle and self-update options.
+- Possibility of having several instances installed at the same time without any conflict.
+
+Under the hood, every command is retrieving the settings from the environment, doing necessary changes using PHP (change folders, create files...), and then calling the appropriate ddev command for the real action: config, start, stop, export-db, import-db, exec, share, composer... The tool abstracts DDEV for most of the users, because they do not need to use the ddev CLI tool for most of the ope ...
```
```suggestion
Under the hood, each command retrieves the settings from the environment, doing necessary changes using PHP (change folders, create files...), and then calling the appropriate DDEV command for the real action: `config`, `start`, `stop`, `export-db`, `import-db`, `exec`, `share`, `composer`... The tool abstracts DDEV for most of the users, because they do not need to use the ddev CLI tool for most of the operations.
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+
+Under the hood, every command is retrieving the settings from the environment, doing necessary changes using PHP (change folders, create files...), and then calling the appropriate ddev command for the real action: config, start, stop, export-db, import-db, exec, share, composer... The tool abstracts DDEV for most of the users, because they do not need to use the ddev CLI tool for most of the ope ...
+
+As an example, consider the steps for installing a shop, like for example [shopware](https://ddev.readthedocs.io/en/latest/users/quickstart/#shopware).
```
```suggestion
For example, consider the steps for installing a shop, like for example [shopware](https://ddev.readthedocs.io/en/latest/users/quickstart/#shopware).
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+Under the hood, every command is retrieving the settings from the environment, doing necessary changes using PHP (change folders, create files...), and then calling the appropriate ddev command for the real action: config, start, stop, export-db, import-db, exec, share, composer... The tool abstracts DDEV for most of the users, because they do not need to use the ddev CLI tool for most of the ope ...
+
+As an example, consider the steps for installing a shop, like for example [shopware](https://ddev.readthedocs.io/en/latest/users/quickstart/#shopware).
+The lpm tool, and due to business needs, has to support all (recent) versions of this shop. And for ensuring that the installation always works the tool needs to set the versions of the dependencies of shopware, like PHP or mariadb, before running the ddev commands, like 'config', for each major version (sometimes, even minor version) of the shop. The user has the need of installing a shop locall ...
```
```suggestion
The lpm tool has to support all (recent) versions of this shop. To ensure that the installation works, the tool needs to set the versions of the dependencies of shopware, like PHP or MariaDB, before running the DDEV commands. The user does not care about how the shop is installed or the differences between different shop versions.
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+As an example, consider the steps for installing a shop, like for example [shopware](https://ddev.readthedocs.io/en/latest/users/quickstart/#shopware).
+The lpm tool, and due to business needs, has to support all (recent) versions of this shop. And for ensuring that the installation always works the tool needs to set the versions of the dependencies of shopware, like PHP or mariadb, before running the ddev commands, like 'config', for each major version (sometimes, even minor version) of the shop. The user has the need of installing a shop locall ...
+
+So, lsm offers to the users a simplified and more focused experience compared to DDEV's experience: DDEV is the general tool that can achieve everything. On the other hand, lsm, powered by DDEV, is the specific tool for doing specific operations. For most of the users, lsm is enough. For most advanced users, lsm is a first step in using DDEV.
```
```suggestion
lsm offers a simplified and more focused experience compared to DDEV's out-of-the-box experience: DDEV is the general tool that can achieve everything. On the other hand, lsm, powered by DDEV, is the specific tool for doing specific operations. For most users, lsm is enough. For most advanced users, lsm is a first step toward using DDEV.
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+
+## The result
+
+It was now much easy to operate with all the shops we support, which are indeed very different each from each other. A user only needed to learn a few commands which are common for all shops. This translated into a notoriously improve in the team's productivity. Many of the problems that, before, may took a lot of time, now they were gone or reduced drastically. This was specially noticeable when ...
```
```suggestion
It is now much easier to operate with all the shops we support, which are indeed very different each from each other. A user only needs to learn a few commands, and they are common for all shops. This translates into significantly improves team productivity. Many of the problems that previously took a lot of time are gone or reduced significantly. This was particularly noticeable when members of the organization needed to use a local shop for a few tests. Using this tool was really much faster than having to deal with the old custom Docker images. Installing a shop locally was no longer a headache.
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+
+It was now much easy to operate with all the shops we support, which are indeed very different each from each other. A user only needed to learn a few commands which are common for all shops. This translated into a notoriously improve in the team's productivity. Many of the problems that, before, may took a lot of time, now they were gone or reduced drastically. This was specially noticeable when ...
+
+The feedback was mostly positive, and feedback also helped to develop lsm further and support more and more shops and features. To this day, there has been 26 releases and almost 400 commits, after a year and a half of internal development.
```
```suggestion
The feedback was mostly positive, and feedback helped to push along lsm's development to support more and more shops and features. There have been 26 releases and almost 400 commits after a year and a half of internal development.
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+
+## In conclusion
+
+Introducing DDEV in a stablish team can be challenging. Some team members may see DDEV as a complex tool to use, with many options and features. Some even may feel lazy about having to learn a new tool, or convinced that the way things are is good enough. Focusing in simplifying the user experience and automating the most common operations can be a game changer and even change the minds of most s ...
```
```suggestion
Introducing DDEV to an establishd team can be challenging. Some team members may see DDEV as a complex tool to use, with many options and features. Some even may feel lazy about having to learn a new tool, or convinced that the way things are is good enough. Focusing on simplifying the user experience and automating the most common operations can be a game changer and can even change the minds of most senior developers in the team.
```

rfay commented on src/content/blog/scripts-powered-by-ddev.md:
```diff
@@ -0,0 +1,78 @@
+## In conclusion
+
+Introducing DDEV in a stablish team can be challenging. Some team members may see DDEV as a complex tool to use, with many options and features. Some even may feel lazy about having to learn a new tool, or convinced that the way things are is good enough. Focusing in simplifying the user experience and automating the most common operations can be a game changer and even change the minds of most s ...
+DDEV's command line interface is flexible enough for allowing more focused tools to be built on top of it.
```
```suggestion
DDEV's command line interface is flexible enough to allow more focused tools to be built on top of it.
```

rfay reviewed (changes requested):
You mention `lpm` and `lsm` tools. Please introduce each and what's different about them, maybe what the initials stand for.

Somewhere you have to say that these are not and will not be publicly available. 

Thanks for taking the time to do this!

(You know that you can just accept suggestions using the GitHub web UI if you agree with them, right?)

agviu commented:
Hi @rfay,
thanks again for your feedback and suggestions. I have added all them and the information you requested.
lpm was a mistake (I had a different tool named like this, but it's not the same as lsm). I replaced it with "lsm".

rfay commented:
This is currently broken, probably because missing featured image. If you have a featured image you'd like to use, say the word. Otherwise I'll try to fix and use ai to generate something. Unfortunately the automated test doesn't show when it's broken this way. Thanks for all the work on this, and we're almost there!

agviu commented:
Hi @rfay , I do not have any featured image. You can use a generated one.
Thanks!

stasadev commented on src/content/blog/extending-ddev-commands-with-scripts.md:
```diff
@@ -0,0 +1,77 @@
+
+Under the hood, every command is retrieving the settings from the environment, doing necessary changes using PHP (change folders, create files...), and then calling the appropriate ddev command for the real action: config, start, stop, export-db, import-db, exec, share, composer... The tool abstracts DDEV for most of the users, because they do not need to use the ddev command-line tool for most o ...
+
+As an example, consider the steps for installing a shop, like for example [shopware](https://ddev.readthedocs.io/en/latest/users/quickstart/#shopware).
```
```suggestion
As an example, consider the steps for installing a shop, like for example [shopware](https://ddev.readthedocs.io/en/stable/users/quickstart/#shopware).
```

rfay commented:
@agviu could you please look at https://20240315-agviu-scripts.ddev-com-front-end.pages.dev/blog/extending-ddev-commands-with-scripts/ and see if it's OK with you?

agviu commented:
@rfay yes it looks very good, thanks!

rfay commented:
Thanks!
